services/mirror/mirror.go

```diff
@@ -96,21 +110,40 @@ func InitSyncMirrors() {
 	if !setting.Mirror.Enabled {
 		return
 	}
-	go graceful.GetManager().RunWithShutdownContext(syncMirrors)
+	mirrorQueue = queue.CreateUniqueQueue("mirror", queueHandle, new(Request))
+
+	go graceful.GetManager().RunWithShutdownFns(mirrorQueue.Run)
 }
 
 // StartToMirror adds repoID to mirror queue
 func StartToMirror(repoID int64) {
 	if !setting.Mirror.Enabled {
 		return
 	}
-	go mirrorQueue.Add(fmt.Sprintf("pull %d", repoID))
+	go func() {
+		err := mirrorQueue.Push(&Request{
+			Type:   PushRequestType,
+			RepoID: repoID,
+		})
+		if err != nil {
+			log.Error("Unable to push push mirror request to the queue for repo[%d]: Error: %v", repoID, err)
+		}
+	}()
 }
 
 // AddPushMirrorToQueue adds the push mirror to the queue
 func AddPushMirrorToQueue(mirrorID int64) {
 	if !setting.Mirror.Enabled {
 		return
 	}
-	go mirrorQueue.Add(fmt.Sprintf("push %d", mirrorID))
+	go func() {
+
+		err := mirrorQueue.Push(&Request{
+			Type:   PullRequestType,
+			RepoID: mirrorID,
+		})
+		if err != nil {
+			log.Error("Unable to push pull mirror request to the queue for repo[%d]: Error: %v", mirrorID, err)
```

Comment:
```suggestion
			log.Error("Unable to add pull mirror request to the queue for repo[%d]: Error: %v", mirrorID, err)
```
Sounds weird otherwise.

Reply:
I'm really not sure about that. I thought about whether add is appropriate - the underlying function is called `Push` not `Add` and the error came from that.

Reply:
Yes, but `unable to push pull mirror` is confusing.